### ISSUE TITLE
AS is required to ignore audience restrictions on claims tokens

### ISSUE THREAD

**Issue by jricher** [KantaraInitiative/wg-uma#160]
When a requesting client submits a token directly to the AS for claims, the AS has absolutely no way of knowing if the requesting client was the intended recipient of that token. Specifically, in OpenID Connect, the `aud` field of the ID Token JWT will contain the `client_id` of the requesting client with the issuer of the JWT. However, the AS is not the issuer of this JWT and therefore has no way of knowing what the `client_id` of the requesting client is with that issuer. The AS cannot differentiate between a valid token and a stolen token, which at least on the surface seems to be a big security hole.

Suggested fix: remove the direct-claims-presentation model until a mechanism that accounts for audience restrictions and token validation can be developed.

**Comment by gffletch** [comment]
Does the OAuth2 token-exchange spec help here?

**Comment by jricher** [comment]
The OAuth2 token-exchange spec doesn't help anywhere.

**Comment by xmlgrrl** [comment]
For reference: The direct submission of a token is the "push" model for a "claims-aware client" defined in Core Sec 3.4.1.2.1 (under https://docs.kantarainitiative.org/uma/rec-uma-core.html#rfc.section.3.4.1.2).

**Comment by mrpotes** [comment]
I agree that this is definitely a problem in the spec as it stands. I think there are two solutions that we can investigate one or both of:

1. The AS is also the IdP, and so is able to verify its own signed id_token, and potentially can choose to verify the UMA Client is the OIDC client to which that token was issued.
2. A trust profile for UMA is written up that describes how a trust relationship might be set up between the AS and issuer of an id_token so that the signature can be verified using certficates from the jwks_uri

**Comment by xmlgrrl** [comment]
Core Sec 6.3 does recommend documenting Claim Token Format Profiles:

https://docs.kantarainitiative.org/uma/rec-uma-core.html#rfc.section.6.3

...and part of that is to "Specify any mutual interpretation details of claim token formats by authorization servers and clients." So one or both of these suggestions could be part of an OpenID Connect profile, for example.

(I wonder if "Claims-Gathering Profile" would be a better name, given that there are other elements besides the mere format that could be standardized. A whole subprotocol and negotiation could go in there, with error_details and such.)

**Comment by xmlgrrl** [comment]
@jricher, you said in the #154 thread that "The client can't reliably push information that the AS can verify in any case that I've seen demonstrated", but I think there's one case (case 1 that @mrpotes alludes to above) where it's technically verifiable and the AS is also not required to ignore the audience restriction (because it issued the client credentials itself). Am I interpreting the situation correctly?

**Comment by jricher** [comment]
Correct -- if you issued the token in the first place you don't need to do anything special with it, since you know everything about the token already. But that's not a particularly interoperable or helpful thing since you can't count on it outside of that one, narrow, highly-optimized case.

**Comment by dom1115** [comment]
Hi,
I'm not an expert of Auth2.0 or UMA, anyway here are my 2 cents:
a) unifying the AS and the the IdP is not a good idea, being exactly contrary to the spirit of UMA to have a clear distinction between the Authentication and Authorization
b) waht about having a proof-of-ownership to the presented the AS with the token?
Some capability based authorization mechanisms envisage such approach both as a way to securely support rights delegation, as well as to assure the requesting party is the actual owner of the presented token.

**Comment by nynymike** [comment]
In some cases, the AS and the OP are the same, so its not true that "the AS is not the issuer of this JWT."

**Comment by jricher** [comment]
But that's only a very limited and optimized case, and it's unnecessary even in that case because the AS/OP can easily have the RqP just log in without telling the client to dangerously spread its ID Tokens around the network.

**Comment by nynymike** [comment]
Personally, I've agreed with this for some time: "remove the direct-claims-presentation model until a mechanism that accounts for audience restrictions and token validation can be developed." 

But I didn't want to ruin it for implementers who had some use cases in mind that I hadn't considered...

**Comment by xmlgrrl** [comment]
I don't know about *very* limited and optimized. We have that use case, and it's not just a one-off. "Easily hav[ing] the RqP just log in" when the client is already trusted by virtue of the ecosystem topology in place is a big burden on those users. What if a claims profile were written to document the trust establishment assumptions in place? That's what they're for.

It was occurring to me that id_tokens would be a lot more useful if they weren't quite so tightly coupled with the surrounding protocol in this way, unlike, say, SAML assertions, which can be (and have been) profiled for use with diverse flows. It all depends on how trust is established. Not all ecosystems need be at risk of having id_tokens dangerously spread around the network.

**Comment by xmlgrrl** [comment]
@dom1115 , colocating an IdP and AS is something we imagined would be a common deployment choice early on, for what it's worth. We don't know how things will unroll in very wide ecosystems yet, but just as a thought experiment, imagine if the current "social IdPs" were to offer authorization services. Then they would be unified IdP/AS services.

**Comment by jricher** [comment]
@xmlgrrl Note that this is colocation of the RqP's IdP and the RO's AS, not the RO's IdP and AS being the same box. This inherently limits it to a single security domain.

**Comment by dom1115** [comment]
@xmlgrrl basing some feature on a "common deployment choice" seems to me not a good solution. From an architectural point of view the IdP and the AS are independent entities, even if they have to trust each other. 
As such they have to properly work independently from their deployment solution (which, BTW, can evolve over time).

**Comment by xmlgrrl** [comment]
Let me address these statements in turn.

Basing a feature on a common deployment choice (another way of saying a common use case) is probably one of the *best* reasons to choose a feature, as long as other less-common but realistic in-scope features are not precluded (a la the OpenID Connect precept of "making simple things simple and complicated things possible" http://openid.net/connect/faq/ ).

That said, what I'm suggesting here is not a "feature" of UMA and certainly not a required feature, but rather a totally optional choice, in fact a profile, the need for which was anticipated by the spec:

https://docs.kantarainitiative.org/uma/rec-uma-core.html#rfc.section.3.4.1.2
"The authorization server has many options for gathering requesting party claims. For example, it could interact with an end-user requesting party directly, or accept claims delivered by a client, or perform a lookup in some external system. The process is extensible and can have dependencies on the type of requesting party (for example, natural person or legal person) or client (for example, browser, native app, or autonomously running web service)."
...
(3.4.1.2.1) "This specification provides a framework for extensibility through claim token format profiling. The authorization server MAY support any number of claim token profiles, and SHOULD document the claim token profiles it supports in its configuration data."

Regarding the independence of the IdP and AS: Indeed they can still be independent, but when they are colocated, opportunities for optimization may arise. The UMA WG has discussed "optimization opportunities" precisely around such colocations (https://groups.google.com/forum/#!searchin/kantara-initiative-uma-wg/optimization$20opportunities), and in fact has built some such opportunities into the spec in its extensibility profiles, e.g., when an AS and RS might be colocated (https://docs.kantarainitiative.org/uma/rec-uma-core.html#rfc.section.5). In other spheres, we can observe the optimization opportunities that have arisen from squishing OpenID and OAuth together (to eliminate two double redirects in the UX) resulting in the flow design that underpins OIDC today.

So, in summary, as long as a separate IdP-and-AS work together just fine, I suggest that it's reasonable for an IdP-that-is-also-an-AS to leverage a claim profile to achieve an optimized flow as long as it doesn't break any rules and is secure and privacy-sensitive.

**Comment by xmlgrrl** [comment]
Discussion on 2015-08-13: Add a security consideration.

**Comment by xmlgrrl** [comment]
As discussed on UMA telecon 2015-08-27: "Looks good; we'll get review of the whole spec later. The last STRONGLY RECOMMENDED should be in active voice, not passive." We can close this, presuming Maciej has already implemented amendments made on the call.
